Sources/NIOHPACK/RingBufferView.swift

```diff
@@ -0,0 +1,101 @@
+//
+//  RingBufferView.swift
+//  NIOHPACK
+//
+//  Created by Jim Dovey on 7/10/18.
+//
+
+import NIO
+
+public struct RingBufferView : ContiguousCollection, RandomAccessCollection {
```

Comment:
Is it worth making this conform to `ContiguousCollection` by forcing a memory copy when we overlap the ring buffer? Or would we do better by just using `Collection` and allowing this to straddle the end of the ring?

Reply:
Possibly, though its main use case is in providing a `ContiguousCollection` to be copied into `ByteBuffer`s. The optimization here (and indeed everywhere else) is to try and avoid the byte-by-byte copy semantics of the `Sequence` writers in `ByteBuffer`.

Reply:
I suspect that's a false economy.

You're trading the byte-by-byte write for two memory allocations and two linear copies, one of the entire ring buffer and one of the view. Even if you assume the byte-by-byte write is 8x slower than a linear copy, and that the view is the entire ring buffer, if the allocations are even 3x more expensive than the copy then the efficiency has vanished. For small copies this is basically guaranteed. For full buffers this is also going to be an issue, as while we do an optimised vectorised copy of ring buffer we still have to copy the whole thing, and if that thing is 8x the size of the view we want then that copy is more expensive than the direct byte-by-byte write.

`ContiguousCollection` is really a protocol that is intended to be applied to things that *already are* contiguous, not to things that can *become* contiguous by allocating. That's why `StaticString` conforms to `ContiguousCollection`, but `String` does not: it's cheaper to use the `Collection` machinery on `String` than to make it contiguous and then `memcpy` the bytes.

Reply:
I've done some tests, and the ratio of copies to direct references is pretty small. Here's the results after running all the decoder tests:

Ring buffer views created: 805842. Number which required copies: 914. Ratio: 0.11342173776000754%

And here for the encoder tests:

Ring buffer views created: 237170. Number which required copies: 873. Ratio: 0.36809039929164733%

Reply:
I think I'd need to see the performance cost of the Collection vs ContiguousCollection over the decoder tests. Mind benchmarking what happens if you change the protocol conformance?